api/service/log_service.go

```diff
@@ -15,53 +15,290 @@
-	logClients map[string]corev1.CoreV1Interface
+	clusterControllers map[string]cluster.Controller
 }
 
 // NewLogService create a log service
-// logClients is a map of cluster name to its cluster client (CoreV1Interface)
-func NewLogService(logClients map[string]corev1.CoreV1Interface) LogService {
-	return &logService{logClients: logClients}
+// clusterControllers is a map of cluster name to its cluster.Controller
+func NewLogService(clusterControllers map[string]cluster.Controller) LogService {
+	color.NoColor = false // To bypasses the check for non-tty output streams.
+
+	return &logService{clusterControllers: clusterControllers}
 }
 
-// ReadLog read log from a container with the given log options
-// Return ReadCloser object which can be used to stream the log when the LogOptions.Follow is set to true
-func (l logService) ReadLog(options *LogQuery) (io.ReadCloser, error) {
-	logClient, ok := l.logClients[options.Cluster]
+func (l logService) StreamLogs(logLineCh chan string, stopCh chan struct{}, options *LogQuery) error {
+	clusterController, ok := l.clusterControllers[options.Cluster]
 	if !ok {
-		return nil, fmt.Errorf("unable to find cluster %s", options.Cluster)
+		return fmt.Errorf("unable to find cluster %s", options.Cluster)
+	}
+
+	namespace := options.Namespace
+	labelSelector := l.getLabelSelector(*options)
+
+	pods, err := clusterController.ListPods(namespace, labelSelector)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		for {
+			select {
+			case <-stopCh:
+				return
+			default:
+			}
+
+			allLogLines := []*LogLine{}
+			mu := &sync.Mutex{}
+
+			var wg sync.WaitGroup
+			for _, pod := range pods.Items {
+				containers := []string{}
+				for _, initContainer := range pod.Spec.InitContainers {
+					containers = append(containers, initContainer.Name)
+				}
+				for _, container := range pod.Spec.Containers {
+					for _, skippedContainer := range skippedContainers {
+						if container.Name == skippedContainer {
+							break
+						}
+
+						containers = append(containers, container.Name)
+					}
+				}
+
+				for _, container := range containers {
+					options.ContainerName = container
+
+					wg.Add(1)
+					go func(clusterController cluster.Controller, namespace string, pod v1.Pod, options LogQuery) {
+						defer wg.Done()
+						logLines := l.getContainerLogs(clusterController, namespace, pod.Name, options)
+
+						mu.Lock()
+						allLogLines = append(allLogLines, logLines...)
+						mu.Unlock()
+					}(clusterController, namespace, pod, *options)
+				}
+			}
+			wg.Wait()
+
+			// Sort all the logs by timestamp in ascending order
+			sort.Slice(allLogLines, func(i, j int) bool {
+				return allLogLines[i].Timestamp.Before(allLogLines[j].Timestamp)
+			})
+
+			for _, logLine := range allLogLines {
+				select {
+				case <-stopCh:
+					return
+				case logLineCh <- logLine.GenerateText(*options):
+				}
+			}
+
+			now := time.Now()
+			options.SinceTime = &now
+
+			time.Sleep(5 * time.Second)
```

Comment:
Do we need to wait this long? It seems too long.

Reply:
I'll reduce it to 1. Previously I increased it for debugging purposes.